crates/loader/src/local/assets.rs

```diff
@@ -129,52 +138,66 @@ fn collect_placement(
-            ),
-            Ok(dir_entry) => {
-                if dir_entry.file_type().is_file() {
-                    let match_path = dir_entry.path();
-                    match to_relative(match_path, &abs) {
-                        Ok(relative_to_match_root_dst) => {
-                            let guest_dst = relative_guest_path.join(relative_to_match_root_dst);
-                            Some(FileMount::from_exact(match_path, &guest_dst))
-                        }
-                        Err(e) => {
-                            let err = Err(e).with_context(|| {
-                                format!(
-                                    "Failed to establish relative path for '{}'",
-                                    match_path.display()
-                                )
-                            });
-                            Some(err)
-                        }
-                    }
-                } else {
-                    None
-                }
-            }
-        })
-        .collect::<Result<Vec<_>>>()?;
+    for de in walker.into_iter() {
+        let dir_entry =
+            de.with_context(|| format!("Failed to walk directory under {}", abs.display()))?;
 
+        if !dir_entry.file_type().is_file() {
+            continue;
+        }
+        let match_path = dir_entry.path();
+        let relative_path = to_relative(match_path, &rel).with_context(|| {
+            format!(
+                "Failed to establish relative path for '{}' relative to '{}'",
+                match_path.display(),
+                rel.as_ref().display(),
+            )
+        })?;
+        if is_excluded_file(Path::new(&relative_path), exclude_patterns) {
+            continue;
+        }
+        let relative_to_match_root_dst = to_relative(match_path, &abs).with_context(|| {
+            format!(
+                "Failed to establish relative path for '{}' relative to '{}'",
+                match_path.display(),
+                abs.display()
+            )
+        })?;
+        let guest_dst = relative_guest_path.join(relative_to_match_root_dst);
+        files.push(FileMount::from_exact(match_path, &guest_dst)?);
+    }
     Ok(files)
 }
 
 /// Generate a vector of file mounts given a file pattern.
-fn collect_pattern(pattern: &str, rel: impl AsRef<Path>) -> Result<Vec<FileMount>> {
+fn collect_pattern(
+    pattern: &str,
+    exclude_patterns: &[glob::Pattern],
+    rel: impl AsRef<Path>,
+) -> Result<Vec<FileMount>> {
     let abs = rel.as_ref().join(pattern);
     log::trace!("Resolving asset file pattern '{:?}'", abs);
 
-    let matches = glob::glob(&abs.to_string_lossy())?;
-    let specifiers = matches
-        .into_iter()
-        .map(|path| FileMount::from(path, &rel))
-        .collect::<Result<Vec<_>>>()?;
-    let files: Vec<_> = specifiers.into_iter().filter(|s| s.src.is_file()).collect();
+    let mut files = Vec::new();
```

Comment:
I update `is_exlcluded_file` to return `Result<bool>`, so probably we can't just use `filter` for it. I separate for loop to multiple iterators so that we can use the `?` operator in each `Result`.